File: src/test/java/diagnostics/DiagnosticsTest.java
package diagnostics;

import com.intellij.psi.PsiErrorElement;
import java.io.File;
import java.io.IOException;
import java.net.URL;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import org.junit.Test;
import org.netbeans.api.project.Project;
import org.netbeans.junit.NbTestCase;
import javaproject.JavaProject;
import org.jetbrains.kotlin.resolve.AnalysisResultWithProvider;
import org.jetbrains.kotlin.resolve.KotlinAnalyzer;
import org.jetbrains.kotlin.utils.ProjectUtils;
import org.jetbrains.kotlin.diagnostics.Diagnostic;
import org.jetbrains.kotlin.diagnostics.Severity;
import org.jetbrains.kotlin.psi.KtFile;
import org.jetbrains.kotlin.resolve.AnalyzingUtils;
import org.openide.filesystems.FileObject;
import static junit.framework.TestCase.assertEquals;
import static junit.framework.TestCase.assertNotNull;
import kotlin.Pair;
import org.netbeans.modules.java.api.common.SourceRoots;
import org.netbeans.modules.java.j2seproject.J2SEProject;
import org.netbeans.modules.java.j2seproject.J2SEProjectGenerator;
import org.netbeans.modules.java.j2seproject.api.J2SEProjectBuilder;
import org.netbeans.modules.project.ant.AntBasedProjectFactorySingleton;
import org.netbeans.spi.project.support.ant.AntProjectHelper;
import org.netbeans.spi.project.support.ant.ProjectGenerator;
import org.openide.filesystems.FileUtil;
import org.w3c.dom.Element;
import org.w3c.dom.NodeList;

/**
 *
 * @author Alexander.Baratynski
 */
public class DiagnosticsTest extends NbTestCase {
    
    private final Project project;
    private final FileObject diagnosticsDir;
    
    public DiagnosticsTest() {
        super("Diagnostics test");
        project = JavaProject.INSTANCE.getJavaProject();
        diagnosticsDir = project.getProjectDirectory().
                getFileObject("src").getFileObject("diagnostics");  
    }
 
    
    private AnalysisResultWithProvider getAnalysisResult(String fileName){
        FileObject fileToAnalyze = diagnosticsDir.getFileObject(fileName);
        
        assertNotNull(fileToAnalyze);
        
        KtFile ktFile = ProjectUtils.getKtFile(fileToAnalyze);
        return KotlinAnalyzer.analyzeFile(project, ktFile);
    }
    
    private void doTest(String fileName, List<Pair<Integer,Integer>> diagnosticsRanges, 
            List<Severity> diagnosticsSeverity, List<Pair<Integer, Integer>> syntaxErrorsRanges) {
        int numberOfDiagnostics = diagnosticsRanges.size();
        int numberOfSyntaxErrors = syntaxErrorsRanges.size();
        
        AnalysisResultWithProvider result = getAnalysisResult(fileName);
        FileObject fileToAnalyze = diagnosticsDir.getFileObject(fileName);
        KtFile ktFile = ProjectUtils.getKtFile(fileToAnalyze);
        
        Collection<Diagnostic> diagnostics = 
                result.getAnalysisResult().getBindingContext().getDiagnostics().all();
        Collection<PsiErrorElement> syntaxErrors = AnalyzingUtils.getSyntaxErrorRanges(ktFile);
        
        assertEquals(numberOfDiagnostics, diagnostics.size());
        assertEquals(numberOfSyntaxErrors, syntaxErrors.size());
        
        if (numberOfDiagnostics > 0) {
            int i = 0;
            for (Diagnostic diagnostic : diagnostics) {
                assertEquals(diagnosticsSeverity.get(i), diagnostic.getSeverity());
                
                Integer startPosition = diagnostic.getTextRanges().get(0).getStartOffset();
                Integer endPosition = diagnostic.getTextRanges().get(0).getEndOffset();
                
                assertEquals(diagnosticsRanges.get(i).getFirst(), startPosition);
                assertEquals(diagnosticsRanges.get(i).getSecond(), endPosition);
                
                i++;
            }
        }
        
        if (numberOfSyntaxErrors > 0) {
            int i = 0;
            for (PsiErrorElement syntaxError : syntaxErrors) {
                Integer startPosition = syntaxError.getTextRange().getStartOffset();
                Integer endPosition = syntaxError.getTextRange().getEndOffset();
        
                assertEquals(syntaxErrorsRanges.get(i).getFirst(), startPosition);
                assertEquals(syntaxErrorsRanges.get(i).getSecond(), endPosition);
                
                i++;
            }
        }
    }
    
    private void doTest(String fileName, List<Pair<Integer,Integer>> diagnosticsRanges, 
            List<Severity> diagnosticsSeverity) {
        doTest(fileName, diagnosticsRanges, diagnosticsSeverity, 
                new ArrayList<Pair<Integer, Integer>>());
    }
    
    private void doTest(String fileName, List<Pair<Integer, Integer>> syntaxErrorsRanges) {
        doTest(fileName, new ArrayList<Pair<Integer, Integer>>(), 
                null, syntaxErrorsRanges);
    }
    
    private void doTest(String fileName) {
        doTest(fileName, new ArrayList<Pair<Integer, Integer>>());
    }
    
    @Test
    public void testProjectCreation() {
        assertNotNull(project);
        assertNotNull(diagnosticsDir);
    }
    
    @Test 
    public void testKtHome() {
        assertNotNull(ProjectUtils.KT_HOME);
    }
    
    @Test
    public void testParameterIsNeverUsedWarning(){
        List<Pair<Integer,Integer>> diagnosticsRanges = 
                new ArrayList<Pair<Integer,Integer>>();
        diagnosticsRanges.add(new Pair<Integer, Integer>(30,33));
        
        List<Severity> severityList = new ArrayList<Severity>();
        severityList.add(Severity.WARNING);
        
        doTest("parameterIsNeverUsed.kt", diagnosticsRanges, severityList);
    }
    
    @Test
    public void testExpectingATopLevelDeclarationError(){
        List<Pair<Integer, Integer>> errorsRanges = 
                new ArrayList<Pair<Integer, Integer>>();
        errorsRanges.add(new Pair<Integer, Integer>(21,31));
        
        doTest("expectingATopLevelDeclaration.kt", errorsRanges);
    }
    
    @Test
    public void testNoTypeMismatch(){
        doTest("checkNoTypeMismatch.kt");
    }
    
    @Test
    public void testTypeMismatch(){
        List<Pair<Integer,Integer>> diagnosticsRanges = 
                new ArrayList<Pair<Integer,Integer>>();
        diagnosticsRanges.add(new Pair<Integer, Integer>(151,172));
        
        List<Severity> severityList = new ArrayList<Severity>();
        severityList.add(Severity.ERROR);
        
        doTest("checkTypeMismatch.kt", diagnosticsRanges, severityList);
    }
    
    @Test
    public void testNoValuePassed(){
        List<Pair<Integer,Integer>> diagnosticsRanges = 
                new ArrayList<Pair<Integer,Integer>>();
        diagnosticsRanges.add(new Pair<Integer, Integer>(80,81));
        
        List<Severity> severityList = new ArrayList<Severity>();
        severityList.add(Severity.ERROR);
        
        doTest("checkNoValuePassed.kt", diagnosticsRanges, severityList);
    }
    
    @Test
    public void testWrongImport(){
        List<Pair<Integer,Integer>> diagnosticsRanges = 
                new ArrayList<Pair<Integer,Integer>>();
        diagnosticsRanges.add(new Pair<Integer, Integer>(28,32));
        
        List<Severity> severityList = new ArrayList<Severity>();
        severityList.add(Severity.ERROR);
        
        doTest("checkWrongImport.kt", diagnosticsRanges, severityList);        
    }
    
    @Test
    public void testCheckNullPointer(){
        List<Pair<Integer,Integer>> diagnosticsRanges = 
                new ArrayList<Pair<Integer,Integer>>();
        diagnosticsRanges.add(new Pair<Integer, Integer>(86,90));
        
        List<Severity> severityList = new ArrayList<Severity>();
        severityList.add(Severity.ERROR);
        
        doTest("checkNullPointer.kt", diagnosticsRanges, severityList);          
    }
    
    @Test
    public void testReassignValue(){
        List<Pair<Integer,Integer>> diagnosticsRanges = 
                new ArrayList<Pair<Integer,Integer>>();
        diagnosticsRanges.add(new Pair<Integer, Integer>(61,66));
        
        List<Severity> severityList = new ArrayList<Severity>();
        severityList.add(Severity.ERROR);
        
        doTest("checkReassignValue.kt", diagnosticsRanges, severityList);         
    }
    
    @Test
    public void testCastIsNeverSucceed(){
        List<Pair<Integer,Integer>> diagnosticsRanges = 
                new ArrayList<Pair<Integer,Integer>>();
        diagnosticsRanges.add(new Pair<Integer, Integer>(95,97));
        diagnosticsRanges.add(new Pair<Integer, Integer>(79,82));
        
        List<Severity> severityList = new ArrayList<Severity>();
        severityList.add(Severity.WARNING);
        severityList.add(Severity.WARNING);
        
        doTest("checkCastIsNeverSucceed.kt", diagnosticsRanges, severityList);        
    }
    
//    public void testTypeParameterOfCollectionDeclaredInJava() {
//        doTest("checkTypeParameterOfCollectionDeclaredInJava.kt");
//    }
    
}

File: src/test/java/diagnostics/DiagnosticsTest.kt
package diagnostics

import utils.*
import com.intellij.psi.PsiErrorElement
import org.netbeans.api.project.Project
import javaproject.JavaProject
import org.jetbrains.kotlin.resolve.AnalysisResultWithProvider
import org.jetbrains.kotlin.resolve.KotlinAnalyzer
import org.jetbrains.kotlin.utils.ProjectUtils
import org.jetbrains.kotlin.diagnostics.Diagnostic
import org.jetbrains.kotlin.diagnostics.Severity
import org.jetbrains.kotlin.psi.KtFile
import org.jetbrains.kotlin.resolve.AnalyzingUtils
import org.openide.filesystems.FileObject

/**
 *
 * @author Alexander.Baratynski
 */
class DiagnosticsTest : KotlinTestCase("Diagnostics test", "diagnostics") {
    
    private fun getAnalysisResult(fileName: String): AnalysisResultWithProvider {
        val fileToAnalyze = dir.getFileObject(fileName)
        assertNotNull(fileToAnalyze)
        
        val ktFile = ProjectUtils.getKtFile(fileToAnalyze)
        return KotlinAnalyzer.analyzeFile(project, ktFile)
    }

    private fun doTest(fileName: String, 
                       diagnosticsRanges: List<Pair<Int, Int>>,
                       diagnosticsSeverity: List<Severity>, 
                       syntaxErrorsRanges: List<Pair<Int, Int>>) {
        val numberOfDiagnostics = diagnosticsRanges.size
        val numberOfSyntaxErrors = syntaxErrorsRanges.size
        
        val result = getAnalysisResult(fileName)
        val fileToAnalyze = dir.getFileObject(fileName)
        val ktFile = ProjectUtils.getKtFile(fileToAnalyze)
        
        val diagnostics = result.analysisResult.bindingContext.diagnostics.all()
        val syntaxErrors = AnalyzingUtils.getSyntaxErrorRanges(ktFile)
        
        assertEquals(numberOfDiagnostics, diagnostics.size)
        assertEquals(numberOfSyntaxErrors, syntaxErrors.size)
        
        diagnostics.forEachIndexed { i, it ->
            assertEquals(diagnosticsSeverity[i], it.severity)
                
            val startPosition = it.textRanges[0].startOffset
            val endPosition = it.textRanges[0].endOffset
                
            assertEquals(diagnosticsRanges[i].first, startPosition)
            assertEquals(diagnosticsRanges[i].second, endPosition)
        }
        
        syntaxErrors.forEachIndexed { i, it ->
            val startPosition = it.textRange.startOffset
            val endPosition = it.textRange.endOffset
                
            assertEquals(syntaxErrorsRanges[i].first, startPosition)
            assertEquals(syntaxErrorsRanges[i].second, endPosition)
        }
        
    }

    private fun doTest(fileName: String, 
                       diagnosticsRanges: List<Pair<Int, Int>>,
                       diagnosticsSeverity: List<Severity>) = doTest(fileName, diagnosticsRanges, diagnosticsSeverity, emptyList())
    
    private fun doTest(fileName: String,
                       syntaxErrorsRanges: List<Pair<Int, Int>> = emptyList()) = doTest(fileName, emptyList(), emptyList(), syntaxErrorsRanges)

    fun testKtHome() = assertNotNull(ProjectUtils.KT_HOME)

    fun testParameterIsNeverUsedWarning() = doTest("parameterIsNeverUsed.kt", listOf(Pair(30, 33)), listOf(Severity.WARNING))

    fun testExpectingATopLevelDeclarationError() = doTest("expectingATopLevelDeclaration.kt", listOf(Pair(21, 31)))

    fun testNoTypeMismatch() = doTest("checkNoTypeMismatch.kt")

    fun testTypeMismatch() = doTest("checkTypeMismatch.kt", listOf(Pair(151, 172)), listOf(Severity.ERROR))

    fun testNoValuePassed() = doTest("checkNoValuePassed.kt", listOf(Pair(80, 81)), listOf(Severity.ERROR))
    
    fun testWrongImport() = doTest("checkWrongImport.kt", listOf(Pair(28, 32)), listOf(Severity.ERROR))

    fun testCheckNullPointer() = doTest("checkNullPointer.kt", listOf(Pair(86, 90)), listOf(Severity.ERROR))

    fun testReassignValue() = doTest("checkReassignValue.kt", listOf(Pair(61, 66)), listOf(Severity.ERROR))

    fun testCastIsNeverSucceed() = doTest("checkCastIsNeverSucceed.kt", listOf(Pair(95, 97), Pair(79, 82)), listOf(Severity.WARNING, Severity.WARNING))

}